# Customer Churn Analysis

## Objective
The objective of this analysis is to identify key factors influencing customer churn and provide actionable business insights to improve customer retention.


## 1. Import Required Libraries

In [1]:
import pandas as pd

## 2. Load Dataset

In [2]:
df = pd.read_csv("Telco Customer Churn.csv")
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 3. Initial Data Exploration
Understanding dataset structure, dimensions, and data types.

In [3]:
df.shape

(7043, 21)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


## 4. Data Cleaning
Handling missing values, correcting data types, and validating dataset integrity.

In [5]:
df[df['TotalCharges'] == " "]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No


In [6]:
df['TotalCharges'].value_counts()

,count
TotalCharges,
,11
20.2,11
19.75,9
20.05,8
19.9,8
...,...
130.15,1
3211.9,1
7843.55,1


In [7]:
df['TotalCharges'] = df['TotalCharges'].replace(" ", pd.NA)

In [8]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'])

In [9]:
df['TotalCharges'] = df['TotalCharges'].fillna(0)

In [10]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [12]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


In [13]:
df.duplicated().sum()

np.int64(0)

### Cleaning Summary
- Replaced blank values in TotalCharges
- Converted TotalCharges to numeric format
- Handled missing values
- Verified no duplicate records
- Dataset is clean and ready for EDA

## 5. Exploratory Data Analysis

In [14]:
df['customerID'].nunique()

7043

### 1. Overall Churn Rate

In [15]:
df['Churn'].value_counts()

,count
Churn,
No,5174
Yes,1869


In [16]:
df['Churn'].value_counts(normalize=True) * 100

,proportion
Churn,
No,73.463013
Yes,26.536987


### Overall Churn Insight

- 26.5% of customers have churned.
- 73.5% of customers are retained.
- The dataset shows moderate class imbalance.
- Customer churn is a significant business concern and requires deeper analysis to identify key drivers.

In [17]:
pd.crosstab(df['Contract'], df['Churn'])

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1307,166
Two year,1647,48


In [18]:
pd.crosstab(df['Contract'], df['Churn'], normalize='index') * 100

Churn,No,Yes
Contract,,
Month-to-month,57.290323,42.709677
One year,88.730482,11.269518
Two year,97.168142,2.831858


### Contract vs Churn Insight

- Month-to-month customers have a very high churn rate (42.7%).
- One-year contract customers show much lower churn (11.2%).
- Two-year contract customers are highly retained (only 2.8% churn).
- Longer contract duration significantly reduces churn risk.

In [19]:
df.groupby('Churn')['tenure'].describe()

,count,mean,std,min,25%,50%,75%,max
Churn,,,,,,,,
No,5174.0,37.569965,24.113777,0.0,15.0,38.0,61.0,72.0
Yes,1869.0,17.979133,19.531123,1.0,2.0,10.0,29.0,72.0


### Tenure vs Churn Insight

- Customers who churn have significantly lower average tenure.
- Retained customers stay much longer with the company.
- Early-stage customers are at higher risk of churn.
- Customer lifetime is strongly associated with churn behavior.

In [20]:
pd.cut(df['tenure'], bins=[0,12,24,48,72]).value_counts()

,count
tenure,
"(48, 72]",2239
"(0, 12]",2175
"(24, 48]",1594
"(12, 24]",1024


In [21]:
pd.crosstab(
    pd.cut(df['tenure'], bins=[0,12,24,48,72]),
    df['Churn'],
    normalize='index'
) * 100

Churn,No,Yes
tenure,,
"(0, 12]",52.321839,47.678161
"(12, 24]",71.289062,28.710938
"(24, 48]",79.611041,20.388959
"(48, 72]",90.486824,9.513176


### Tenure Category vs Churn

- Customers in the first 12 months show the highest churn rate.
- Churn rate consistently decreases as tenure increases.
- Long-term customers (48+ months) are highly retained.
- Early customer engagement is critical to reduce churn.

In [22]:
pd.crosstab(df['TechSupport'], df['Churn'], normalize='index') * 100

Churn,No,Yes
TechSupport,,
No,58.364526,41.635474
No internet service,92.595020,7.404980
Yes,84.833659,15.166341


### Tech Support Impact

- Customers without tech support have significantly higher churn.
- Customers who use tech support services are more likely to stay.
- Providing strong customer support reduces churn risk.

In [23]:
pd.crosstab(df['InternetService'], df['Churn'], normalize='index') * 100

Churn,No,Yes
InternetService,,
DSL,81.040892,18.959108
Fiber optic,58.107235,41.892765
No,92.595020,7.404980


### Internet Service Impact

- Fiber optic users show higher churn compared to DSL users.
- Customers without internet service have the lowest churn.
- Service type influences customer retention.

In [24]:
pd.crosstab(df['OnlineSecurity'], df['Churn'], normalize='index') * 100

Churn,No,Yes
OnlineSecurity,,
No,58.233276,41.766724
No internet service,92.595020,7.404980
Yes,85.388806,14.611194


### Online Security Impact

- Customers without online security are more likely to churn.
- Value-added services increase retention.

In [25]:
df.groupby('Churn')['MonthlyCharges'].mean()

,MonthlyCharges
Churn,
No,61.265124
Yes,74.441332


### Monthly Charges Insight

- Customers who churn pay higher average monthly charges.
- High pricing may increase churn risk.
- Pricing strategy should be evaluated.

In [26]:
pd.crosstab(df['PaymentMethod'], df['Churn'], normalize='index') * 100

Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),83.290155,16.709845
Credit card (automatic),84.756899,15.243101
Electronic check,54.714588,45.285412
Mailed check,80.893300,19.106700


### Payment Method Impact

- Customers using electronic check have the highest churn.
- Automatic payment methods show lower churn rates.
- Encouraging auto-pay may improve retention.

In [27]:
df.to_csv("telco_churn_cleaned.csv", index=False)

## 6. Project Conclusion

- Data cleaning and validation completed successfully.
- Exploratory Data Analysis identified key churn drivers:
  - Contract type
  - Tenure duration
  - Tech support availability
  - Internet service type
  - Payment method
  - Monthly charges
- Early-stage customers and month-to-month subscribers show highest churn risk.
- Service quality and pricing strategy play a significant role in retention.
- Clean dataset exported for SQL analysis and dashboard development.

The dataset is now ready for business intelligence reporting.